### PR TITLE
Add saptune discovery

### PR DESCRIPTION
**Pull request by rtorrero**
Add initial implementation of the `saptune --format json status` and expose this through the collector

### CODE REVIEW

File: internal/agent/agent.go
Comment (arbulu89) [inline internal/agent/agent.go:50]: Is this `InstanceName` used for something? I didn't see any usage in the discovery struct

File: internal/core/saptune/saptune.go
Comment (arbulu89) [inline internal/core/saptune/saptune.go]: Do we need all this structs here?
I mean, we are just bypassing the json, we don't use them at all internally

File: internal/core/saptune/saptune.go
Comment (arbulu89) [inline internal/core/saptune/saptune.go:21]: Most probably you want to revisit this to use the new format with `package_version`, `saptune_installed`, etc

File: internal/core/saptune/saptune.go
Comment (arbulu89) [inline internal/core/saptune/saptune.go:75]: Let's wrap the coming `err` here, otherwise we just get the generic error, without any context of why it failed

File: internal/core/saptune/saptune.go
Comment (arbulu89) [inline internal/core/saptune/saptune.go:72]: Strange to print this here if we are not going to print as `Info` the version afterwards

File: internal/core/saptune/saptune.go
Comment (arbulu89) [inline internal/core/saptune/saptune.go:84]: Let's remove this initial empty line

File: internal/core/saptune/saptune.go
Comment (arbulu89) [inline internal/core/saptune/saptune.go:102]: Be careful with this. Saptune has some predefined exit codes, from 0 to 4 i think, where they still return the correct format, but it is determined if saptune tuning is working or not

File: internal/core/saptune/saptune.go
Comment (arbulu89) [inline internal/core/saptune/saptune.go:109]: Do we need to unmarshall, why don't we send directly the json output in the discovery payload?

**Review by arbulu89 — COMMENTED** [review]
Hey @rtorrero ,
Some initial comments.
Besides that, I wonder how this code could be reusable in a gatherer. I would maybe expect some `RunSaptuneCommand(cmd, format)`, where you pass the command you want to execute and maybe the format.

File: internal/core/saptune/saptune.go
Comment (rtorrero) [reply internal/core/saptune/saptune.go]: No need and can be removed. I thought they might be used in the future if we need to do something with the JSON output before sending it, but I can remove & re-add when needed.

File: internal/core/saptune/saptune.go
Comment (arbulu89) [inline internal/core/saptune/saptune.go:56]: These initial log messages are outdated, right?

File: internal/core/saptune/saptune.go
Comment (arbulu89) [inline internal/core/saptune/saptune.go:62]: I don't think we should do any error handling here.
We should return output and err, directly.
As commented, in `saptune status`, having an error code doesn't mean that the command failed.
You can debug the output before returning

File: internal/core/saptune/saptune.go
Comment (arbulu89) [inline internal/core/saptune/saptune.go:52]: Are we sure we want to run this for every command?
If we had the `isJsonSupported` field, we could avoid running this. 
Actually, the code owner (us) could even avoid using `RunCommand`, as we know that json is not supported before hand

File: internal/core/saptune/saptune.go
Comment (arbulu89) [inline internal/core/saptune/saptune.go:20]: I don't think we need the json annotation.
This struct is internal, not meant to be marshalled

File: internal/core/saptune/saptune_test.go
Comment (arbulu89) [inline internal/core/saptune/saptune_test.go:19]: I'm a bit confused with the tests.
I would like to have some tests for the `NewSaptune` and other tests for the `RunCommand`.
They don't need to be combined. For the `RunCommand` you can create the saptune struct of your convinience

File: internal/discovery/saptune.go
Comment (arbulu89) [inline internal/discovery/saptune.go:45]: We cannot do that. Even if we get some error here, we need to send a payload, with the `PackageVersion: null/empty string`, `SaptuneInstalled: false` and `Status: nil`

File: internal/discovery/saptune.go
Comment (arbulu89) [inline internal/discovery/saptune.go:53]: At this point `SaptuneInstalled` should be always `true`, right?
I shouldn't need a comparison

File: internal/discovery/saptune.go
Comment (arbulu89) [inline internal/discovery/saptune.go:49]: We cannot ignore the error.
What if json format is not supported?
That's why I like to have the `IsJsonSupported` field.
Because you can check it here, and if it is false, send a payload with `status: nil` without even using `RunCommand`

**Review by arbulu89 — CHANGES_REQUESTED** [review]
Hey @rtorrero 
Some feedback.
Consider adding a `IsJsonSupported` field. I think it will come really useful in the gatherers as well, as we can skip running command functions since the beginning

File: internal/discovery/saptune.go
Comment (arbulu89) [reply internal/discovery/saptune.go:49]: Alright, I understood now.
You are redirecting the nil coming from the `RunCommand`, a bit ugly for me, but it should work

File: internal/core/saptune/saptune.go
Comment (arbulu89) [inline internal/core/saptune/saptune.go:46]: I think you could simply return `Saptune{}`

File: internal/core/saptune/saptune.go
Comment (arbulu89) [inline internal/core/saptune/saptune.go:49]: Maybe better to store the result in a variable, just to improve readability

File: internal/core/saptune/saptune.go
Comment (arbulu89) [inline internal/core/saptune/saptune.go:53]: If we forbid non json options, we should hardcode the `--format json` in the command, so the user doesn't need to put it.

**Review by arbulu89 — COMMENTED** [review]
Looking good!
Almost ready to merge, once we fix the linting things mostly

File: internal/core/saptune/saptune_test.go
Comment (arbulu89) [inline internal/core/saptune/saptune_test.go:28]: Maybe we could name this `expectedSaptune`. Wdyt?

File: internal/core/saptune/saptune.go
Comment (arbulu89) [inline internal/core/saptune/saptune.go:76]: We could reuse the main `RunCommand` once we have the new args.
Not a big deal

File: internal/discovery/saptune.go
Comment (arbulu89) [inline internal/discovery/saptune.go:64]: I think we should capture the error and exit if there is any.
Just for safetiness. 
The status should always return  a json, but just in case...

**Review by arbulu89 — APPROVED** [review]
@rtorrero green light from my side.
It would be nice to run some manual integration test between agent and web to see that it works.
(remember that there is a pending PR in web which fixes the policy)

File: internal/discovery/saptune.go
Comment (rtorrero) [reply internal/discovery/saptune.go:64]: The thing is that as you said, some `saptune` commands will return non-exit codes which still have valid data. If we exit we wouldn't return that JSON, right? or do you mean to simply:

```
		if err != nil {
			log.Debugf("Error while unmarshalling saptune status: %s", err)
		}
```
and thats it?

**Pull request merged**
